### ISSUE TITLE
SimpleMessageListenerContainer should be offer more aggressive scale down options [AMQP-745]

### ISSUE THREAD

**Issue by spring-operator** [spring-projects/spring-amqp#2305]
**[Yannick](https://jira.spring.io/secure/ViewProfile.jspa?name=yenki)** opened **[AMQP-745](https://jira.spring.io/browse/AMQP-745?redirect=false)** and commented

This is somehow related to #2259

Unless we set 'consecutiveIdleTrigger' to 1 and an ridiculously low receiveTimeout, it's very hard to have consumers removed from a queue having reasonable amount of activity given the current SimpleMessageListenerContainer implementation.

For instance I have a queue with 100 msg/s and a consumer handling 1 msg each millisecond. This means that one consumer should easily be able to sustain the load all by itself.
Now even if I set 'consecutiveIdleTrigger' to 1, I could theoretically have 100 consumers and none of them would be removed. That make it 1000 If I were to use the default 'consecutiveIdleTrigger' value of 10.

#2259 sounds like a very good way to solve the problem but more short term I think having a base class with template and/or overridable methods would do the trick for us to hook and inject a more aggressive scale down strategy. I can contribute something I just don't know exactly which solution is more likely to be merged upstream. Please advise.

---

**Affects:** 1.6.8

**Reference URL:** #2259

**Comment by TatianaNeuer** [comment]
Hi, is there any news on this ? 

We are facing a similar behaviour: SimpleMessageListenerContainer creates way more consumers than needed. Instead, we would like the number of consumers to adjust according to the load. So I tried a few things.

Even if 1 consumer should handle all the load by itself, SimpleMessageListenerContainer quickly adds more and more consumers until it reaches maxConcurrentConsumers.
=> I tried limiting the number of consumers by counting busy consumers (consumers that have consecutiveMessages >= consecutiveActiveTrigger) and allowing up to twice the number of busy consumers (and up to maxConcurrentConsumers). That worked for limiting the number of created consumers.

When the load on the queue decreases, the number of consumers does not scale down.
=> I tried adding consumer priorities (first consumer would be set with a priority of 100, second consumer would be set with a priority of 99, and so on). The load was unbalanced between consumers, leaving some consumers with no messages at all so they were removed.

This solution adjusts the number of consumers to suit the load, however with a constant load the number of consumers fluctuates quite rapidly so a smoother adjustment would certainly be better. 

Gary Russell's idea seems smoother but configuring a desired rate per consumer means knowing in advance how many messages can be handled by one consumer, what if the consumers suddenly need more time to process messages ?

I can contribute but I don't know what solution would be the best. Can you please advise ?

**Comment by artembilan** [comment]
The logic there is like this:
```
		private void checkAdjust(boolean receivedOk) {
			if (receivedOk) {
				if (isActive(this.consumer)) {
					this.consecutiveIdles = 0;
					if (this.consecutiveMessages++ > SimpleMessageListenerContainer.this.consecutiveActiveTrigger) {
						considerAddingAConsumer();
						this.consecutiveMessages = 0;
					}
				}
			}
			else {
				this.consecutiveMessages = 0;
				if (this.consecutiveIdles++ > SimpleMessageListenerContainer.this.consecutiveIdleTrigger) {
					considerStoppingAConsumer(this.consumer);
					this.consecutiveIdles = 0;
				}
			}
		}
```
So, if the current consumer receives the message we start a new consumer, but only if the `consecutiveMessages` does not exceed that `consecutiveActiveTrigger`.
Otherwise we stop it when its `consecutiveIdles` exceeds the `consecutiveIdleTrigger`.
Does not those `consecutiveActiveTrigger` & `consecutiveIdleTrigger` setting to `1` help?
```
	/**
	 * If {@link #maxConcurrentConsumers} is greater then {@link #concurrentConsumers}, and
	 * {@link #maxConcurrentConsumers} has not been reached, specifies the number of
	 * consecutive cycles when a single consumer was active, in order to consider
	 * starting a new consumer. If the consumer goes idle for one cycle, the counter is reset.
	 * This is impacted by the {@link #batchSize}.
	 * Default is 10 consecutive messages.
	 * @param consecutiveActiveTrigger The number of consecutive receives to trigger a new consumer.
	 * @see #setMaxConcurrentConsumers(int)
	 * @see #setStartConsumerMinInterval(long)
	 * @see #setBatchSize(int)
	 */
	public final void setConsecutiveActiveTrigger(int consecutiveActiveTrigger) {
		Assert.isTrue(consecutiveActiveTrigger > 0, "'consecutiveActiveTrigger' must be > 0");
		this.consecutiveActiveTrigger = consecutiveActiveTrigger;
	}

	/**
	 * If {@link #maxConcurrentConsumers} is greater then {@link #concurrentConsumers}, and
	 * the number of consumers exceeds {@link #concurrentConsumers}, specifies the
	 * number of consecutive receive attempts that return no data; after which we consider
	 * stopping a consumer. The idle time is effectively
	 * {@link #receiveTimeout} * {@link #batchSize} * this value because the consumer thread waits for
	 * a message for up to {@link #receiveTimeout} up to {@link #batchSize} times.
	 * Default is 10 consecutive idles.
	 * @param consecutiveIdleTrigger The number of consecutive timeouts to trigger stopping a consumer.
	 * @see #setMaxConcurrentConsumers(int)
	 * @see #setStopConsumerMinInterval(long)
	 * @see #setReceiveTimeout(long)
	 * @see #setBatchSize(int)
	 */
	public final void setConsecutiveIdleTrigger(int consecutiveIdleTrigger) {
		Assert.isTrue(consecutiveIdleTrigger > 0, "'consecutiveIdleTrigger' must be > 0");
		this.consecutiveIdleTrigger = consecutiveIdleTrigger;
	}
```

**Comment by artembilan** [comment]
Another piece of logic is like this:
```
				boolean receivedOk = receiveAndExecute(this.consumer); // At least one message received
				if (SimpleMessageListenerContainer.this.maxConcurrentConsumers != null) {
					checkAdjust(receivedOk);
				}
```
So, if this `maxConcurrentConsumers` is not set, then on new dynamic consumers are created.

**Comment by TatianaNeuer** [comment]
Thank you for your answer.

I tried setting `consecutiveIdleTrigger` to 1 and setting `consecutiveActiveTrigger` to 10000, it created consumers at a much slower pace, however the number of consumers still reached `maxConcurrentConsumers` and never decreased.

In my case, setting `consecutiveIdleTrigger` to 1 and `receiveTimeout` to 5 made the number of consumers automatically adjust to the workload.
Why is `receiveTimeout` set to 1000 by default ? is it safe to set it to a much lower value ?

**Comment by artembilan** [comment]
Well, the behavior with the `maxConcurrentConsumers` is correct. It point is to increase to that number and keep them.
That's something similar to the  `concurrentConsumers` where all of them are started eagerly immediately.

The `receiveTimeout` is used to block a consumer for polling from the in-memory queue in the `BlockingQueueConsumer`:
```
Message message = handle(this.queue.poll(timeout, TimeUnit.MILLISECONDS));
```
So, it really might be safe to set it into a smaller period, but not by default.

We don't see a problem having those concurrent consumers on the queue if they are idling.
It does not effect RabbitMQ performance.
Or do I miss something?

Thank you for sharing your experience!

**Comment by artembilan** [comment]
The number of opened channels can be limited by the `CachingConnectionFactory`, which is `25` by default.

> While with some workloads this is a natural state of the system, long lived channels should be used instead when possible.

And that's exactly what is going on with that `CachingConnectionFactory`: it does keep only that number of channels per connection:
```
	/**
	 * Sets the channel checkout timeout. When greater than 0, enables channel limiting
	 * in that the {@link #channelCacheSize} becomes the total number of available channels per
	 * connection rather than a simple cache size. Note that changing the {@link #channelCacheSize}
	 * does not affect the limit on existing connection(s), invoke {@link #destroy()} to cause a
	 * new connection to be created with the new limit.
	 * <p>
	 * Since 1.5.5, also applies to getting a connection when the cache mode is CONNECTION.
	 * @param channelCheckoutTimeout the timeout in milliseconds; default 0 (channel limiting not enabled).
	 * @since 1.4.2
	 * @see #setConnectionLimit(int)
	 */
	public void setChannelCheckoutTimeout(long channelCheckoutTimeout) {
```

I'm sorry for saying that, but Gary Russell is not with our team any more and I'm not sure (yet) what exactly you would like to see as a fix for your requirement.
I'm opened for any suggestions, but for now I don't see a way how to make it smart enough for scaling down when load is lower.

Might be this part can be improved somehow with an `else`:
```
				if (isActive(this.consumer)) {
					this.consecutiveIdles = 0;
					if (this.consecutiveMessages++ > SimpleMessageListenerContainer.this.consecutiveActiveTrigger) {
						considerAddingAConsumer();
						this.consecutiveMessages = 0;
					}
				}
```

I mean if `consecutiveMessages` is bigger than `consecutiveActiveTrigger` we add consumer, otherwise `considerStoppingAConsumer()`.

WDYT?

**Comment by joyfulviper** [comment]
Hello,

I would like to inquire if there are ongoing discussions regarding this issue or if there is a specific method to immediately scale down a consumer in SimpleMessageListenerContainer (SMLC).

Context & Logic I am implementing a logic where specific methods have different concurrency constraints within a single application:

method1(): cannot run in parallel with another method1().

method2(): can run in parallel with method2() or method1().

To achieve this without using explicit application-level locks (and to leverage the RabbitMQ queue for buffering), I am dynamically adjusting the container's concurrency within the listener. The goal is to process method1 with a single consumer, then scale up to allow another consumer to pick up the next message while the first executes method2.

Here is the pseudocode of my implementation:

```java
@RabbitListener(
        id = "#{config.getId()}",
        group = "group1",
        queues = "#{config.getQueueNameList()}",
        ackMode = "#{T(org.springframework.amqp.core.AcknowledgeMode).MANUAL}",
        concurrency = "1",
        autoStartup = "false"
)
public void listen(Message message, Channel channel) throws Exception {
  try {
    var properties = message.getMessageProperties();
    
    // 1. Critical Section (Concurrency: 1)
    var result = method1(message);
    
    // 2. Scale UP: Allow next message to be consumed by a new thread
    var container = (SimpleMessageListenerContainer) registry.getListenerContainer(config.getQueueNameList());
    container.setConcurrentConsumers(2);
    
    // 3. Non-blocking Section
    method2(result);
    
    channel.basicAck(properties.getDeliveryTag(), false);    
  } catch(Exception e) {
    channel.basicNack(properties.getDeliveryTag(), true);
  }  finally {
    // 4. Scale DOWN: Attempt to reduce consumers back to 1
    var container = (SimpleMessageListenerContainer) registry.getListenerContainer(config.getQueueNameList());
    container.setConcurrentConsumers(1);
  }
}
```
The Problem setConcurrentConsumers(n) works perfectly when scaling up (increasing consumers). However, when scaling down (decreasing), the reduction is not immediate. I have tried tuning consecutiveIdleTrigger, receiveTimeout, and maxConcurrentConsumers, but the active consumer does not terminate immediately after processing the current message.

Desired Behavior I want the current consumer thread to switch to a "cancel" state and terminate immediately after the current message is processed (and acked), ensuring it does not attempt to fetch or wait for the next message.

Observations

`DirectMessageListenerContainer`: I considered this container, but it is designed to map consumers to queues (1:1 or parallel based on queue list size), which conflicts with my requirement for dynamic concurrency on a single queue.

`channel.basicCancel()`: I tried calling `channel.basicCancel(properties.getConsumerTag())` directly. While this successfully prevents the current consumer from fetching the next message, it effectively "kills" the consumer without a clean mechanism to revive or restart it within the container's management lifecycle.
Question If there is no existing option to support immediate scale-down in SMLC:

Is there a recommended workaround for this pattern?

If I implement a feature to support "immediate scale-down after current processing," would this be considered a valuable contribution to the project?

I am willing to contribute code if this is a valid requirement others might share.

Thank you.

**Comment by joyfulviper** [comment]
Hi, I apologize for the very late follow-up — it's been about 5 months since my last comment.

I've been experimenting with a prototype to solve the immediate scale-down problem I described, and I'd like to propose an approach before submitting a PR.

**Problem recap**

When `setConcurrentConsumers()` is called from within a listener to scale down, `adjustConsumers()` picks an arbitrary consumer from the `HashSet` iterator to cancel:

```java
// SimpleMessageListenerContainer.adjustConsumers()
Iterator<BlockingQueueConsumer> consumerIterator = this.consumers.iterator();
while (consumerIterator.hasNext() && delta > 0 ...) {
BlockingQueueConsumer consumer = consumerIterator.next(); // arbitrary pick
    consumer.basicCancel(true);
    consumerIterator.remove();
delta--;
        }
```

This can cancel a consumer that is still actively processing a message, while the consumer that just finished picks up the next message — defeating the purpose of the scale-down.

**Proposed solution**

Add an `aggressiveScaleDown` flag to `SimpleMessageListenerContainer`:

1. **`setConcurrentConsumers()` scale-down behavior**: When `aggressiveScaleDown` is enabled, scale-down skips `adjustConsumers()` and only updates the field. Scale-up still works as before.

```java
// in setConcurrentConsumers()
if (isActive()) {
    if (this.aggressiveScaleDown && delta > 0) {
        // defer to checkAdjust — the current consumer will stop itself
    } else {
        adjustConsumers(delta);
    }
}
```

2. **`checkAdjust()` stops the current consumer**: After `receiveAndExecute()` returns, `checkAdjust()` detects excess consumers and the consumer that just finished processing stops **itself** — not a random one.

```java
// in AsyncMessageProcessingConsumer.checkAdjust()
if (aggressiveScaleDown && consumers.size() > concurrentConsumers) {
    this.consumer.basicCancel(true);
    consumers.remove(this.consumer);  // stops THIS consumer, not arbitrary
    this.markedForStop = true;
}
```

3. **`markedForStop` flag to exit the main loop immediately**: Without this, the broker can deliver a message between `commitIfNecessary` (ack) and `basicCancel`, causing `hasDelivery()` to stay `true` and the consumer to process one more message.

```java
// before (consumer drains prefetched messages even after cancel)
while (isActive(consumer) || consumer.hasDelivery() || !consumer.cancelled()) {

// after (exits immediately when marked for stop)
while (!this.markedForStop && (isActive(consumer) || consumer.hasDelivery() || !consumer.cancelled())) {
```

4. **Requires `maxConcurrentConsumers` to be set**: This aligns with the existing design where `checkAdjust()` is only invoked when `maxConcurrentConsumers != null`.

**Interaction with existing properties**

When `aggressiveScaleDown` is enabled, scale-down related properties are bypassed since the intent is immediate reaction:
- `consecutiveIdleTrigger` — bypassed, excess consumers stop immediately without waiting for idle cycles
- `stopConsumerMinInterval` — bypassed, no minimum interval between consumer stops

Scale-up related properties remain unaffected:
- `consecutiveActiveTrigger` — still controls when `considerAddingAConsumer()` is called
- `startConsumerMinInterval` — still enforces minimum interval between new consumer starts

**Usage example**
```java
// Factory configuration
factory.setConcurrentConsumers(1);
factory.setMaxConcurrentConsumers(2);
factory.setAggressiveScaleDown(true);

// Inside listener
container.setConcurrentConsumers(2); // scale up — immediate
container.setConcurrentConsumers(1); // scale down — deferred to checkAdjust
```

Would this approach be acceptable for a contribution? I'm happy to adjust the design based on your feedback before submitting a PR.

**Comment by artembilan** [comment]
Hi, @joyfulviper !

Thank you for such a great analysis and proposal.

I don't see a problem implementing your feature.
Only probably "aggressive" is too aggressive word, along side with "kill", "abort", "execute" or "back" and "white" in the end.
How about to call that option like `immediateScaleDown`?
Looks like that is exactly how you explain it in your proposal:

> since the intent is immediate reaction

I say go for PR and we may incorporate it into upcoming `4.1.0-RC1` next Monday.

**Comment by joyfulviper** [comment]
Thanks for the suggestion! I've renamed it to immediateScaleDown. Submitting the PR now.
